twenty-first/src/util_types/algebraic_prng.rs

```diff
@@ -0,0 +1,37 @@
+use itertools::Itertools;
+
+use crate::shared_math::x_field_element::XFieldElement;
+
+use super::algebraic_hasher::SpongeHasher;
+
+pub trait AlgebraicPrng {
+    type PrngState;
+    fn sample_xfield(state: &mut Self::PrngState, num_elements: usize) -> Vec<XFieldElement>;
+    fn sample_indices(state: &mut Self::PrngState, max: u32, num_elements: usize) -> Vec<u32>;
+}
+
+impl<H: SpongeHasher> AlgebraicPrng for H {
+    type PrngState = H::SpongeState;
+
+    fn sample_xfield(state: &mut Self::PrngState, num_elements: usize) -> Vec<XFieldElement> {
+        let num_squeezes =
+            (num_elements * 3 + <H as SpongeHasher>::RATE - 1) / <H as SpongeHasher>::RATE;
+        (0..num_squeezes)
+            .into_iter()
+            .flat_map(|_| <H as SpongeHasher>::squeeze(state))
+            .collect_vec()
+            .chunks(3)
+            .map(|ch| XFieldElement::new(ch.try_into().unwrap()))
+            .collect_vec()
+    }
+
+    fn sample_indices(state: &mut Self::PrngState, max: u32, num_indices: usize) -> Vec<u32> {
```

Comment:
If `sample_indices()` does not require that `max` is a power of 2, we will lack an algorithm for sampling indices on Triton VM. I assume that our only use-case of sample indices that aren't powers of 2 will be deprecated (it's in the current FRI, but it's not necessary).

Reply:
I would assume that `sample_indices(sponge, max, num_indices).len() == num_indices`, but the current implementation rounds up to the nearest multiple of `RATE`. While this is equivalent to the behavior on Triton VM with the `squeeze` instruction, I'm leaning towards preferring actually getting `num_indices` values out.